### ISSUE TITLE
New syntax for configuration of property setters

### ISSUE THREAD

**Issue by dskalski** [FakeItEasy/FakeItEasy#175]
I have simple class
```c#
public class Simple
{
    public virtual int VirtualProperty { get; set; }
}
```
When i run (FakeItEasy.1.13.1)
```c#
var strict = A.Fake<Simple>(options => options.Strict());
A.CallTo(() => strict.VirtualProperty).CallsBaseMethod();
strict.VirtualProperty = 999;
```
I get an error
Call to non configured method "set_VirtualProperty" of strict fake.

And I have to
```c#
var strict = A.Fake<Simple>(options => options.Strict());
A.CallTo(strict).Where(a => a.Method.Name == "get_VirtualProperty").CallsBaseMethod();
A.CallTo(strict).Where(a => a.Method.Name == "set_VirtualProperty").CallsBaseMethod();
strict.VirtualProperty = 999;
```
Does CallBaseMethod () works for virtual property? What am I doing wrong

---
*Update*

The proposed syntax is:

```C#
A.CallToSet(() => foo.Bar)…
A.CallToSet(() => foo.Bar).To(42)…
A.CallToSet(() => foo.Bar["the answer"]).To(42)…
A.CallToSet(() => foo.Bar).To(() => A<int>.That.IsGreaterThan(0)))…
A.CallToSet(() => …).WithAnyArguments()…
A.CallToSet(() => …).WhenArgumentsMatch(call => …)…
```

These methods would allow configuration of the call in the same way that is currently provided by

```C#
A.CallTo(foo).Where(a => a.Method.Name == "set_Foo" ... )...
```

**Comment by adamralph** [comment]
Are we seeing any light at the end of the tunnel for this?

**Comment by blairconrad** [comment]
Oh, man. I just reread. I still wish we had a better syntax, but am not sure that anything has come along that helps.
Unless we can Roslyn it up, but that's currently beyond me…

**Comment by adamralph** [comment]
I don't see how Roslyn can help...

**Comment by blairconrad** [comment]
That's probably because I don't yet understand how Roslyn works!
I figured we'd be able to use the compiler as a service to notice when we had a set action (as a lambda, not an expression) inside an `A.CallTo` and, using the fancy fancy syntax tree inspection powers, rewrite the darn thing so it was something that our current rule builders would understand, maybe even directly to `A.CallTo(fake).Where(x => x.Method.Name == "set_Foo")` and so on.
I should go read some Roslyn things again. I watched a video on how to make inspections, but the details remain fuzzy.

**Comment by thomaslevesque** [comment]
Roslyn can only help if we have the source code, but we don't... the closest we can get is an `Expression`, and they're quite limited.

**Comment by adamralph** [comment]
@blairconrad I don't think what you are suggesting is possible. Effectively, we'd be extended the compiler and swapping in our own version of it, which I don't think is really an intended use case, nor one that I'd advise, since we'd then effectively be saying that you use "FakeItEasy-flavoured C#" to write your tests.

**Comment by thomaslevesque** [comment]
One thing that we *could* do is to use the same approach as [Shouldly](https://github.com/shouldly/shouldly). It takes advantage of debug symbols to look at the original source code, which allows it to produce more informative messages.

**Comment by blairconrad** [comment]
@adamralph, oh, if we were gong to do that, I'd suggest that FakeItEasy-flavoured C# still be valid C#. We wouldn't want to work on anything other than an AST that was valid. It's just that we could potentially take an A.CallTo that contains a func such as `A.CallTo(() => fake.Prop = 17)` and interrogate it to figure out which proper was being set on which fake.
I agree that it may not be practical.

Oh, @thomaslevesque, I had wondered how Shouldly did that. How does it work in release mode with no PDBs?

**Comment by adamralph** [comment]
@blairconrad ah, I see what you mean. Yes, I guess we could load the assembly in via Roslyn and play around with the AST to try and find that out.

**Comment by adamralph** [comment]
I second @blairconrad's question about release mode. Many project prefer to run their CI, including tests, in release mode.

**Comment by thomaslevesque** [comment]
> Oh, @thomaslevesque, I had wondered how Shouldly did that. How does it work in release mode with no PDBs?

It doesn't, if the PDBs aren't there. You get a less informative failed assertion message, along with a hint that the full debug information is necessary for more detailed messages.

However, you can build in Release mode with full debug info (just for the test projects), so it's not really an issue.

**Comment by adamralph** [comment]
> However, you can build in Release mode with full debug info (just for the test projects), so it's not really an issue.

You'd more than likely have to build the SUT projects with full debug info, since that's where the interesting types would be defined.

**Comment by thomaslevesque** [comment]
> You'd more than likely have to build the SUT projects with full debug info, since that's where the interesting types would be defined.

As far as i can tell, Shouldly only examines the unit test code... If you write this in your test:

     foo.Bar.ShouldBe(42);

Shouldly only needs the `foo.Bar` part, and even then, it only needs it as text.

**Comment by adamralph** [comment]
What about

```C#
myActualDomainObject.ShouldBe(myExpectedDomainObject):
```

?

**Comment by thomaslevesque** [comment]
What about it? It just needs to extract "myActualDomainObject" from the unit test source file...

**Comment by adamralph** [comment]
The `MyDomainType` class would be defined in the SUT assembly, not the unit test assembly.

No problem, I think I misunderstood. This is not about doing funky things with the SUT, it's about verbalising the subject of the assertion, i.e. the `foo.Bar`. I think I get it now :wink:

**Comment by blairconrad** [comment]
In #637, @thomaslevesque said he liked

```c#
A.CallToSetter(() => fake.SimpleProperty)
A.CallToSetter(() => fake.SimpleProperty).WithValue(42)
A.CallToSetter(() => fake.IndexedProperty).WithIndex("foo")
```

Me too. But it only works if `fake.SimpleProperty` has a getter.

**Comment by adamralph** [comment]
Another option is to do something similar to NSubstitute, http://stackoverflow.com/questions/7637024/nsubstitute-intercepting-setter-only-property-invocation

I.e. we configure the setter of the fake by setting it to something like the magical `Arg.Do<T>()` which performs some action when the SUT sets the property.

It would be a syntax not rooted in A, but we could probably make it work.

**Comment by adamralph** [comment]
More info on `Arg.Do<T>()` http://nsubstitute.github.io/help/actions-with-arguments/#performing_actions_with_arguments

**Comment by blairconrad** [comment]
@adamralph, keep in mind that the magical `Arg.Do<T>()` will have to be an actual `T` in order to satisfy the compiler. I'm not sure how NSubstitute gets around that without ambient state and junk. I mean, they must be, but…

**Comment by blairconrad** [comment]
```c#
public static T Do<T>(Action<T> useArgument)
{
    return ArgSpecQueue.EnqueueSpecFor<T>(new AnyArgumentMatcher(typeof(T)), x => useArgument((T)x));
}
```

I can't take any more time on this now, but this smells of ambient state to me. We'd have to structure a call very carefully to avoid.

**Comment by adamralph** [comment]
@blairconrad I believe you're right. There's no handle to the fake there and no guarantee that `Do` is called on the RHS of a property assignment.

**Comment by thomaslevesque** [comment]
> Me too. But it only works if fake.SimpleProperty has a getter.

True. But I don't think I saw more than 2 or 3 write-only properties in almost 15 years of C# programming... (and one of them was in FakeItEasy :wink:). This is really a corner case, and for those, it's always possible to use `A.CallTo(fake).Where(call => call.Method.Name == "set_TheProperty")`. _Or_, we could have an overload that accepts the property name:

    A.CallToSetter(fake, nameof(fake.SimpleProperty))

> Another option is to do something similar to NSubstitute, http://stackoverflow.com/questions/7637024/nsubstitute-intercepting-setter-only-property-invocation

[I'm not a huge fan](https://xkcd.com/1378/) (thanks @blairconrad for the link :wink:). I don't like the fact that it's a completely different approach from all other call specifications.

**Comment by thomaslevesque** [comment]
I realized there was a mistake in my indexed property example. This wouldn't compile:

    A.CallToSetter(() => fake.IndexedProperty).WithIndex("foo")

because it's not legal to refer to the property with no index. The correct syntax would be:

    A.CallToSetter(() => fake.IndexedProperty["foo"])

(which looks much better anyway)

And if we don't care about the index value:

    A.CallToSetter(() => fake.IndexedProperty[A<string>._])

**Comment by adamralph** [comment]
> I don't think I saw more than 2 or 3 write-only properties in almost 15 years of C# programming... 

It's a valid point. I wouldn't be too bothered about requiring people to pass the name of the property for in this case.

**Comment by adamralph** [comment]
I think I prefer the equivalent English of

```C#
A.CallToSet(() => foo.Bar)
```

or

```C#
A.CallToSetProperty(() => foo.Bar)
```

(from https://github.com/FakeItEasy/FakeItEasy/issues/175#issuecomment-25745811)

**Comment by thomaslevesque** [comment]
`A.CallToSet` is fine by me. `A.CallToSetProperty` is probably more verbose than necessary.

**Comment by adamralph** [comment]
And to specify an argument constraint:

```C#
A.CallToSet(() => foo.Bar).To(42)
```

**Comment by blairconrad** [comment]
... and indexes?

**Comment by thomaslevesque** [comment]
> And to specify an argument constraint:
> 
>     A.CallToSet(() => foo.Bar).To(42)

Yes, it's probably more natural than `WithValue`.

> ... and indexes?

Is it not covered here? https://github.com/FakeItEasy/FakeItEasy/issues/175#issuecomment-205852628

**Comment by adamralph** [comment]
```C#
A.CallToSet(() => foo.Bar["the answer"]]).To(42)
```

**Comment by blairconrad** [comment]
Sorry, @thomaslevesque, I didn't scroll back far enough.

So, the plan is to rely on there being a getter with the same indexes as the setter. The innards of 
`A.CallToSet(() => foo.Bar["the answer"]]).To(42)` would be an expression, so we can interrogate the indices in the usual manner.
We can omit the `To` in cases where we don't care what the value is to be, or supply one for "equals"-style matching.

Am I caught up? We'd just be missing

1. a way to configure setters that don't have a signature-matching getter, and
2. fancy discrimination on the value supplied to the setter (e.g. "when the value is greater than 7")

Not criticizing, just wondering if I've got it all.

**Comment by thomaslevesque** [comment]
> a way to configure setters that don't have a signature-matching getter,

I don't think you _can_ have a getter and setter with different signatures... the indexes are defined at the property level, not at the getter and setter level. So that leaves only the issue of write-only properties.



> fancy discrimination on the value supplied to the setter (e.g. "when the value is greater than 7")

It could also be supported:

    A.CallToSet(() => foo.Bar).To(() => A<int>.That.IsGreaterThan(0)));

(the lambda is a little annoying, but it wouldn't work without it...)

**Comment by adamralph** [comment]
> 1. a way to configure setters that don't have a signature-matching getter

As @thomaslevesque says, this is a subset of the setter-only case. As above, we can rely on the reflection based method for this corner case, e.g. `A.CallTo(f).Where(x => x.Method.Name == ... && Method.Argments... )`

> 2. fancy discrimination on the value supplied to the setter (e.g. "when the value is greater than 7")

@thomaslevesque beat me to it, I had the same thing in mind. It's a shame about the lambda but I don't see any other way to specify argument constraints, given that the value can't appear in the lambda supplied to `A.CallToSet()`. It really is such a shame that "An expression tree may not contain an assignment operator". :disappointed:

**Comment by thomaslevesque** [comment]
> It's a shame about the lambda

What worries me is that it's very easy to forget the lambda and write this:

    A.CallToSet(() => foo.Bar).To(A<int>.That.IsGreaterThan(0)));

[Oh, wait](https://github.com/FakeItEasy/FakeItEasy/issues/559)... We could also add a diagnostic for this in the analyzer.

> It really is such a shame that "An expression tree may not contain an assignment operator". :disappointed:

Yes... none of the features added after C# 3 are supported, I think.

**Comment by adamralph** [comment]
Excellent. It sounds like we have a plan... @blairconrad can you pick any holes?

**Comment by blairconrad** [comment]
> I don't think you can have a getter and setter with different signatures

er, yes. I guess a signature mismatch means "no getter". :disappointed: 

And just to make sure I understand the value-specification, we're looking at

```c#
A.CallToSet(Expression<Func<TValue>>).To(TValue expected).DoesNothing(); // matches if the provided value Equals expected
A.CallToSet(Expression<Func<TValue>>).To(Expression<Func<TValue>> expectedMatcher).DoesNothing(); // matches like arguments do in normal A.CallTos
A.CallToSet(Expression<Func<TValue>>).DoesNothing(); // convenience overload for ignoring value type
```

If the lambda is forgotten in the second case, the invocation of the matcher will throw an exception, and the analyzer may warn about this as well (less urgent, IMO). If the `Expression<Func<TValue>>` isn't a property getter then we throw. Or if there's no setter that matches the getter.

We have compile-time type matching on the argument to `A.CallToSet()`, and for the type provided by `To`, which is good.

The result of `To` will probaby be a `IVoidArgumentValidationConfiguration` so nobody can specify returns on it, so no mistakes there…

If anyone has a setter that doesn't have an accessible getter, then they are on their own with the powerful `A.CallTo(object)`. Tough luck for them.

I think it sounds good.

I like the syntax, and I think the limitations are worth it if it means we can get rid of `NextCall.To`.

**Comment by adamralph** [comment]
I do believe we are all systems go... :astonished:

**Comment by blairconrad** [comment]
Oh! My comment made it through! That's super exciting. Now all we need is someone to implement it before I finish #637. Because otherwise, I'll feel compelled to make `NextCall.To` work, and then remove that implementation.

Any takers?

**Comment by adamralph** [comment]
Not right now, but I'd love to when I have time. In the meantime, I'm happy for anyone else to take it.

**Comment by Alxandr** [comment]
Could there be a string overload for the set only case? It's quite nice now that we have nameof in the language.

**Comment by thomaslevesque** [comment]
> Could there be a string overload for the set only case? It's quite nice now that we have nameof in the language.

We could probably add it; actually I think we should.

**Comment by blairconrad** [comment]
What kind of API are we considering? And is this growing too big?
I've added a sample (somewhat) friendly entry point below:

```c#
A.CallToSet(Expression<Func<TValue>>).To(TValue expected).DoesNothing(); // matches if the provided value Equals expected
A.CallToSet(Expression<Func<TValue>>).To(Expression<Func<TValue>> expectedMatcher).DoesNothing(); // matches like arguments do in normal A.CallTos
A.CallToSet(Expression<Func<TValue>>).DoesNothing(); // convenience overload for ignoring value type

A.CallToSet(string).On(object).WithIndices(Expression<Func<object[]>).To(Expression<Func<object>>).DoesNothing();
// and overloads could omit WithIndices and/or To
// and overloads for WithIndices(object[]) and To(object)
```

Once we're specifying the property by string, we're losing type-safety anyhow. Maybe instead once one indicates the fake and the name of the property, we just allow additional configuration via a `IWhereConfiguration<IAnyCallConfigurationWithNoReturnTypeSpecified>`-like thing? (Maybe that thing.)

**Comment by blairconrad** [comment]
> Not right now, but I'd love to when I have time. In the meantime, I'm happy for anyone else to take it.

I don't mind doing it, but since I'm the long lead person anyhow, it'll push out 2.0 more if I take it.
Still, it's fine with me if nobody else wants it. Just say the word or assign me or something.

**Comment by blairconrad** [comment]
Darn it. I had a thought.

Instead of 

```c#

**Comment by thomaslevesque** [comment]
> What kind of API are we considering? And is this growing too big?

Ah, I had only considered the `A.CallToSet(string)` part. What you propose to check the value and indices would work, but I wonder if it's not getting too complex to be worth it... The benefit over a plain `A.CallTo(fake).Where(...)` isn't compelling IMO.

**Comment by adamralph** [comment]
To prevent scope creep, I propose we lean on `A.CallTo(fake).Where(...)` for now, and spin off the string overload into a separate issue, probably post 2.0.

**Comment by thomaslevesque** [comment]
> To prevent scope creep, I propose we lean on `A.CallTo(fake).Where(...)` for now, and spin off the string overload into a separate issue, probably post 2.0.

:+1:

**Comment by blairconrad** [comment]
I'm content as well, as it seems an edge case. It does mean that @Alxandr and his ilk will be initially discomfited and then will have to make an additional change if they want the fancy version.
Although really, they'd probably just quick make up their own convenience extension not unlike this:

```c#
public static class SetOnlyPropertyExtension
{
    public static IAnyCallConfigurationWithNoReturnTypeSpecified SettingProperty(this IAnyCallConfigurationWithNoReturnTypeSpecified config, string propertyName)
    {
        return config.Where(call => call.Method.Name == "set_" + propertyName);
    }
}
```

and then call something like

```c#
A.CallTo(setOnly).SettingProperty(nameof(ISetOnly.MyProperty)).MustHaveHappened();
```

So I doubt it'd be much of a bother for them.

**Comment by blairconrad** [comment]
Hey, @thomaslevesque, now that I think about it, how do you feel about `A.CallToSet`'s Intellisense behaviour?

In unrelated news, I'm far enough along with #637, that I might as well pick up the implementation of this issue, to save me the trouble of mucking about with the `NextCall.To`.
Unless nobody minds `NextCall.To` going away before we implement this change? We don't have to _release_ the package without `NextCall`.

But if I'm to implement this, #642, and #637, I might as well do them in the "proper" order. :-)

**Comment by adamralph** [comment]
@blairconrad sure, please go ahead and take this. Thanks!

**Comment by thomaslevesque** [comment]
> Hey, @thomaslevesque, now that I think about it, how do you feel about A.CallToSet's Intellisense behaviour?

Didn't notice this comment until now... I don't think it will interfere with Intellisense; it should come right after `A.CallTo`, which is fine

**Comment by adamralph** [comment]
> it should come right after A.CallTo, which is fine

Agreed.

**Comment by blairconrad** [comment]
small wrinkle, even assuming work continues well:

I'd thought to return an `IVoidArgumentValidationConfiguration` from `A.CallToSet`. This interface will allow callers to further specify many action, as well as the restriction `WithAnyArguments` and `WhenArgumentsMatch`, which is fine (unless you think it isn't), except that a simple implementation would return a `RuleBuilder` object that will consider _all_ the arguments to the `set_PropertyName` method, where the last "argument" is the value to which the property is set.
(An `Invokes` or `Throws` would see all the arguments in the exact same way.)

This would provide clients the same interface as when configuring a property setter using `A.CallTo(fake).Where(call => call.Method.Name == 'set_PropertyName)`. 

Is this what we want? It's probably the most convenient way to go, but would require some explanation for users of `WhenArgumentsMatch`, `Throws`, and `Invokes`. (I'm not so worried about `WithAnyArguments`, as a subsequent `To` could override the restriction on the "last argument".)

An alternative would be to return a custom interface, which could provide property-centric versions of `WhenArgumentsMatch` (although we'd probably be better off with something else, like `Matching`, as clients may want to base the specification on all the indexes as well as the new value) and `Invokes`.
This may make more sense to clients, except for the discrepancy with `A.CallTo(fake).Where(…)`, but will be quite a lot more work.

**Comment by thomaslevesque** [comment]
> I'd thought to return an `IVoidArgumentValidationConfiguration` from `A.CallToSet`.

I was thinking about a more specific interface (`IPropertySetterConfiguration` or something similar), that would perhaps inherit from `IVoidArgumentValidationConfiguration`. Otherwise you can't have `.To(...)` or `.WithIndices(...)`.

> This would provide clients the same interface as when configuring a property setter using `A.CallTo(fake).Where(call => call.Method.Name == 'set_PropertyName)`.

I don't think it's a problem, except that we might want to hide `WhenArgumentsMatch` in Intellisense, because it's a bit tricky to use with properties (mostly in the case of indexed properties).

**Comment by blairconrad** [comment]
>  a more specific interface 

Yes. I shouldn't type at night. I had really been thinking of `IVoidArgumentValidationConfiguration` + what we need to support `To`.

`WithIndices` had been suggested as part of the syntax we'd considered for configuring write-only properties, which I thought had been deferred until post-2.0.

> we might want to hide `WhenArgumentsMatch` in Intellisense,

You think hiding is enough? If we're going to cobble together a new interface with `To` anyhow, we could just exclude it altogether. Hmm. It may not matter. I think this discussion has given me enough material to continue to experiment with. Thanks, @thomaslevesque!

**Comment by thomaslevesque** [comment]
> If we're going to cobble together a new interface with To anyhow, we could just exclude it altogether.

Well, I was thinking about making that interface inherit from `IVoidArgumentValidationConfiguration`, but if it's doesn't, then we don't need to hide it.

**Comment by blairconrad** [comment]
I think it's cleaner to just not expose the member at all, but I'll keep the "hide it from Intellisense" option in my back pocket, just in case.

Right now, my biggest concern is converting an expression that accesses a property getter into one that accesses a property setter. But I will plug away!

**Comment by adamralph** [comment]
@blairconrad @thomaslevesque 

I've added an update to the issue description with the proposed syntax. Can you please check to see if I got it right?

**Comment by blairconrad** [comment]
@adamralph, @thomaslevesque, we'd also proposed (and I think it's still reasonable) to have an overload that omits the `.To`. It's really a shortcut for `To(() => A<int>._` or similar, but may be useful.

As for allowing configuration in the same way that `A.CallTo(foo).Where(a => a.Method.Name == "set_Foo" … ) …` goes, I think we should support a replacement for `WhenArgumentsMatch` that would allow the caller to supply a predicate that could consider all of the indices in conjunction with the value that was set. Likewise, a replacement for `WithAnyArguments`. So perhaps


```C#
A.CallToSet(() => foo.Bar)…
A.CallToSet(() => foo.Bar).To(42)…
A.CallToSet(() => foo.Bar["the answer"]).To(42)…
A.CallToSet(() => foo.Bar).To(() => A<int>.That.IsGreaterThan(0)))…
A.CallToSet(() => …).WithAnyIndices()… // or with the .To as well
A.CallToSet(() => …).Where(call => …)… // or with the .To as well
```

Where the argument to `Where` would contain a collection for the indices as well as the supplied value.

I'm still of two minds about the `WithAnyIndices` and the `Where`, though. I think they are the right thing to do if we want to fully support property setter configuration using helpful language.
However, they will then conflict with the way we use `Throws` and `Invokes`, which only know about all of the _arguments_ to the method (the indices and the value). It seems to me that we should either

1. fully commit, and supply custom versions of `Throws` and `Invokes` as well, or
2, keep the existing methods and also relay on `WithAnyArguments` and `WhenArgumentsMatch` instead of `WithAnyIndices` and `Where`

**Comment by thomaslevesque** [comment]
> I've added an update to the issue description with the proposed syntax. Can you please check to see if I got it right?

Looks good to me, with the same remark as @blairconrad about omitting the `.To(...)`.

@blairconrad, how do you envision the signature of `Where`? The current `Where` takes a predicate on an `IFakeObjectCall`; it would have to be something different for a property assignment if you want to expose the value and indices separately.

> that would allow the caller to supply a predicate that could consider all of the indices in conjunction with the value that was set

Is this really useful? I can't come up with a realistic scenario where it would be necessary.

**Comment by patrik-hagne** [comment]
I'm in a bar, so this might be far off... Is there a way to figure out or
"guess" that the user is after the setter using the current syntax with
"CallTo" and "WhereArgumentsMatch"? I kind of cringe at the thought of
introducing a "special case" method directly on "A".

Cheers!

Den mån 18 apr. 2016 20:18Thomas Levesque <notifications@github.com> skrev:

**Comment by blairconrad** [comment]
@thomaslevesque, yes, `Where` would have to take a predicate that acted new interface that exposed the indices and value.
The whole thing seems like quite a lot of work and fairly large surface area to cover what I fear may be a little-used feature. (And by that I mean configuration of property setters.)

> > that would allow the caller to supply a predicate that could consider all of the indices in conjunction with the value that was set

> Is this really useful? I can't come up with a realistic scenario where it would be necessary.

I don't know. I'd probably never use it. On the other hand, saying it's not useful is equivalent to saying that there's no need to let clients inspect the final argument to a call when using the traditional method-based `Where`.

Maybe we want the `Bonus` property on the `Compensation` object to throw an exception when index `Programmer` is assigned a value greater than `0.10`, but not when index `Manager` is. :wink:

**Comment by blairconrad** [comment]
@patrik-hagne, howdy.

You're suggesting using `A.CallTo` with an expression containing the property getter still, and relying on detection inside `WhenArgumentsMatch` to determine that we're talking about the setter?
I'd be surprised if that's possible, since `WhenArgumentsMatch` would already be working on a getter specification.

I suppose an alternative would've been a special case method on `IVoidConfiguration`, but that's been [talked about before](https://github.com/FakeItEasy/FakeItEasy/issues/175#issuecomment-25745811).

**Comment by thomaslevesque** [comment]
> The whole thing seems like quite a lot of work and fairly large surface area to cover what I fear may be a little-used feature. (And by that I mean configuration of property setters.)

> [...]

> I don't know. I'd probably never use it. On the other hand, saying it's not useful is equivalent to saying that there's no need to let clients inspect the final argument to a call when using the traditional method-based Where.

Well, for those rare cases, there's always `A.CallTo(fake).Where(...)`... As long as this workaround exist, I think it's OK not to support every corner case in the "user-friendly" API.

> I'm in a bar, so this might be far off... Is there a way to figure out or "guess" that the user is after the setter using the current syntax with "CallTo" and "WhereArgumentsMatch"? I kind of cringe at the thought of introducing a "special case" method directly on "A".

I agree that it's not ideal, but there doesn't seem to be any good way around it...

**Comment by adamralph** [comment]
I'm not convinced it's worth providing custom property setter versions of all the various config methods. To clarify, is it true to say that all the various scenarios would be covered by the current methods, albeit in a slightly less than savoury syntax?

In response to @patrik-hagne, we managed to avoid the adding of a specialised config method for property setters to `A` for a long time but, with [the removal of `NextCall.To`](https://github.com/FakeItEasy/FakeItEasy/issues/642) we felt we needed something better than the string based reflection of `A.CallTo(fake).Where(...)` to offer for this scenario. It's not ideal, but `A.CallTo()` will still appear before it in intellisense, and, ultimately, having it available *somewhere* off `A` is probably a good thing.

**Comment by adamralph** [comment]
I've updated the description with @blairconrad's more comprehensive syntax examples (using the current methods for now). Feel free to edit if required.

**Comment by thomaslevesque** [comment]
> To clarify, is it true to say that all the various scenarios would be covered by the current methods, albeit in a slightly less than savoury syntax?

I think so. Is there _anything_ you can't specify with `A.CallTo(fake).Where(...)`? Sure, the syntax is ugly, but it's very expressive.

**Comment by blairconrad** [comment]
@adamralph, I think the examples look good. I'll know more when I play with a working codebase. On that note,
https://github.com/blairconrad/FakeItEasy/tree/new-property-setter-syntax has a minimal working implementation of the simplest syntax:

```c#
[Scenario]
public static void ConfiguringSetWithAnyValue(
    IHaveInterestingProperties subject,
    bool wasConfiguredBehaviorUsed)
{
    "Given a Fake with a configurable property"
        .x(() => subject = A.Fake<IHaveInterestingProperties>());

    "And I configure a property setter irrespective of assigned value"
        .x(() => A.CallToSet(() => subject.ThrowingProperty).Invokes(call => wasConfiguredBehaviorUsed = true));

    "When I call the property setter"
        .x(() => subject.ThrowingProperty = 7);

    "Then the configured behavior will be used"
        .x(() => wasConfiguredBehaviorUsed.Should().BeTrue());
}
```

And also a number of invalid attempts at configuration are detected and properly complained about!

**Comment by blairconrad** [comment]
Those spec step names leave something to be desired. "Given/And I"?

**Comment by thomaslevesque** [comment]
> Those spec step names leave something to be desired. "Given/And I"?

I don't mind about "Given/And I", but the property name is misleading.

**Comment by adamralph** [comment]
How about:

```gherkin
Given a fake with a property # I think "configurable" can be assumed
  And assignment of the property is configured for any value
 When I assign the property
 Then the configured behaviour is used
```

**Comment by blairconrad** [comment]
> the property name is misleading.

@thomaslevesque, that's a good point. 

@adamralph, thanks!

**Comment by blairconrad** [comment]
Gents, 

> A.CallToSet(() => …).WithAnyArguments()… // or with the .To as well
> A.CallToSet(() => …).WhenArgumentsMatch(call => …)… // or with the .To as well

It occurs to me that there's no point in having the fluent API support using `To` followed by these options. They will trounce whatever expectation was set in the `To`. Shall we just drop the combination from the API? That is, once `To` has been specified, we'd return a `IVoidConfiguration` instead of a `IVoidArgumentValidationConfiguration`?

**Comment by adamralph** [comment]
:+1: makes sense to me.

**Comment by blairconrad** [comment]
Thanks. Nearly no impact on the code, but the documentation becomes easier. I'll update the description upstairs.

**Comment by thomaslevesque** [comment]
> That is, once To has been specified, we'd return a IVoidConfiguration instead of a IVoidArgumentValidationConfiguration.

:+1:

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).
